### ISSUE TITLE
[REQUEST]: Identify duplicate demangled symbols

### ISSUE THREAD

**Issue by Coder-256** [compiler-explorer/compiler-explorer#6255]
### Is your feature request related to a problem? Please describe

In Rust, you can have multiple different symbols with the same demangled result. For example, this comes up when using closure-to-fn coercion: Rust will generate different implementations of `FnOnce::call_once` for each closure, but they all demangle to the same thing (`core::ops::function::FnOnce::call_once`).

For example:

```rust
pub fn get_adder() -> fn(u32, u32) -> u32 {
    |x, y| { x + y }
}

pub fn get_multiplier() -> fn(u32, u32) -> u32 {
    |x, y| { x * y }
}
```

This code results in [the following assembly](https://rust.godbolt.org/z/oedxrh4x6) when "demangle identifiers" is disabled:

```
_ZN4core3ops8function6FnOnce9call_once17h1d230b22b4f5ee7aE:
        lea     eax, [rdi + rsi]
        ret

_ZN4core3ops8function6FnOnce9call_once17h8b0b7f7b4815ab50E:
        mov     eax, edi
        imul    eax, esi
        ret

_ZN7example9get_adder17hcbc33e165d3b6d9dE:
        mov     eax, offset _ZN4core3ops8function6FnOnce9call_once17h1d230b22b4f5ee7aE
        ret

_ZN7example14get_multiplier17h7782378fdbbffd30E:
        mov     eax, offset _ZN4core3ops8function6FnOnce9call_once17h8b0b7f7b4815ab50E
        ret
```

However, with "demangle identifiers" enabled, the two closures seem to have the same name, and `get_adder` and `get_multiplier` appear to be identical:

```
core::ops::function::FnOnce::call_once:
        lea     eax, [rdi + rsi]
        ret

core::ops::function::FnOnce::call_once:
        mov     eax, edi
        imul    eax, esi
        ret

example::get_adder:
        mov     eax, offset core::ops::function::FnOnce::call_once
        ret

example::get_multiplier:
        mov     eax, offset core::ops::function::FnOnce::call_once
        ret
```

### Describe the solution you'd like

Perhaps append a counter at the end of functions appearing with the same demangled name. For example, something along the lines of: `core::ops::function::FnOnce::call_once`, `core::ops::function::FnOnce::call_once{1}`, `core::ops::function::FnOnce::call_once{2}`, etc.

### Describe alternatives you've considered

None

### Additional context

Not applicable

**Comment by narpfel** [comment]
`rustc` has two name mangling schemes, `legacy` (the default for now) and `v0` (which will become the default at some point in the future).

`legacy` disambiguates duplicate symbol names (such as the `call_once` methods of different closures) by including a unique-per-symbol hash in each name. CE’s demangler could include this hash, but doesn’t currently. FWIW, `c++filt` includes the hash by default.

`v0` was designed so that a demangled name should be unique by itself. (Note: It still uses a hash to disambiguate crates with the same name.)

Always including the hash might make sense for `legacy` mangling, but for `v0` it’s mostly unnecessary information (duplicate crates in different versions are rare on CE, also the `v0` RFC [recommends omitting the crate-id](https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html#appendix-a---suggested-demangling)), so I’d suggest a new checkbox *Options...* → *Include disambiguating hashes in identifiers* (wording TBD) that controls whether hashes are included. Another wording could be *Verbose demangling*, which is what `c++filt` calls this.

`v0` demanglings for a closure’s `call_once` method without and with crate-ids:
```
<example::get_multiplier::{closure#0} as core::ops::function::FnOnce<(u32, u32)>>::call_once:
<example[e8d519bf5b4bcf09]::get_multiplier::{closure#0} as core[5a612027cd243909]::ops::function::FnOnce<(u32, u32)>>::call_once:
```

As a workaround, `-C symbol-mangling-version=v0` can be passed to the compiler: https://godbolt.org/z/zxEf7P6rj

**Comment by partouf** [comment]
Using v0 if available seems to make a lot more sense to me.
It's not easy to add something like a counter as it will require a lot more logic, or we would have to make them unique by adding the hash to the demangled name.

(for reference, demangler is here https://github.com/compiler-explorer/tools/blob/main/rust/src/main.rs (we might need to rebuild this anyway) - but the names are passed 1 by 1 by this TS code https://github.com/compiler-explorer/compiler-explorer/tree/main/lib/demangler)

**Comment by narpfel** [comment]
I implemented the checkbox I proposed in #6265.

Also, this issue has come up before: #1754.
